### ISSUE TITLE
EPUB3 `<aside>` footnotes rendered twice

### ISSUE THREAD

**Issue by Jules-Bertholet** [koreader/koreader#8623]
* KOReader version: 2021.12.1
* Device: All

#### Issue
EPUB3 `<aside epub:type="footnote">` footnotes are rendered twice (once inline, and once at the bottom of the page).
In contrast, other EPUB readers, like [epub.js](https://github.com/futurepress/epub.js/), only render these once (at the bottom of the page or as a pop-up).

This means that, for books that store their footnotes interspersed within the main content in the HTML, the inline footnotes interrupt the flow of text rendered in KOReader.

In addition, KOReader doesn't allow selecting text or clicking on links of the bottom-of-page footnotes.

#### Steps to reproduce
1. Open the [attached sample EPUB](https://github.com/koreader/koreader/files/7823608/footnote.zip) in KOReader (had to change file extension to ZIP to appease GitHub, you will need to change it back)
2. Navigate to the second page
3. Observe that the footnote is displayed twice, and that the one at the bottom of the page can't be interacted with
![KoReader Footnote](https://user-images.githubusercontent.com/79955405/148424391-7f7bd06c-c660-42b2-9883-cf787487766c.png)

**Comment by poire-z** [comment]
That's perfectly normal, the way crengine works, and I feel it is also the safest option.
I explained how it works in various past dicussions, so just linking to some of them I found:

https://github.com/koreader/koreader/issues/6622#issuecomment-685686730 https://github.com/koreader/koreader/issues/6622#issuecomment-691446819 - these also answers your:
> In addition, KOReader doesn't allow selecting text or clicking on links of the bottom-of-page footnotes.

https://github.com/koreader/crengine/issues/446#issuecomment-883297413
I think I described all this even more precisely, with more comments about footnotes somewhere else that I can't find.
Anyway, a summary:
- we render the whole book without any notion of footnotes, including the sections with the footnotes, which have to be rendered for the next step to work. Some footnote section with CSS `-cr-hint: footnote-inpage` are remembered as candidates for inclusion at the bottom of pages.
- when making pages by slicing the whold document height, we can find out footnotes slices that could be cut/copied and pasted at the bottom of the page, making the real page shorter, and pasting the footnote at the place kept at the bottom: it's really just an image (the coordinates are off, and may be composed of many bits, that's why no text selection or click, too complex to solve, nor worth the complexifying).
- but the original footnote content is still part of the book, and must exist and be visible for the above to work.

There is one solution for EPUB makers to solve this: it's to have footnote content in EPUB fragments marked "non-linear", and we have support for masking these section: #6809 #6847
That has to be done by the publisher, we can't do it ourselves, as we could have footnote content mixed with any other valuable content, and we would need to build real page from multiple height slices, which look unsafe, risky, and complex.

Your screenshot shows the worst case as it's a small and single fragment fitting all on the page, so you have a real page content of 30px height, that makes a page - and then the footnote content is put at the bottom of the page, so you indeed see it twice. Usually, it's at end of book. Sometimes, it's at end of the chapter, which is a bit bothering, but these are then shorter and easier to skip with a few taps.

> In contrast, other EPUB readers, like epub.js, only render these once (at the bottom of the page or as a pop-up).

Do they do that only when the footnotes are in a `<aside>` ?
We don't process ASIDE in any particular way. An idea would be to consider ASIDE as non-linear fragments then, but that's not obvious.

TLDR: small issue, quite hard to address, but bearable :)

**Comment by Jules-Bertholet** [comment]
> Do they do that only when the footnotes are in a <aside> ?

Yes, that is how it works. For example, Apple iBooks is documented to behave this way: https://help.apple.com/itc/booksassetguide/en.lproj/itccf8ecf5c8.html

Processing these as non-linear fragments seems like it would work, if it's not too difficult to implement...

**Comment by poire-z** [comment]
Btw, can you check on your real book(s) that you noticed this with that it has not already non-linear fragments (Top menu, first tab, `[ ] Hide non-linear fragments`) ?
Because on the 3 first books that I found with `<aside type=footnote>` in, these aside were all wrapped in a xhtml fragment marked as non-linear, ie:
![image](https://user-images.githubusercontent.com/24273478/148447344-dd29cb6f-d3b2-4ea6-8b2d-3e696102d6c1.png)
(I'm not saying standalone ASIDE does not need hiding, but if publishers are used to wrap them in a xhtml marked as non-linear in the .opf, the need is reduced.)

And it is each footnote that is wrapped in a ASIDE, so in a book with 1000 footnotes, that makes 1000 non-linear flows + 1 linear flow, so 1001 things for the frontend, and it would look ugly on the UI side (and possibly slow).

**Comment by Jules-Bertholet** [comment]
@poire-z The `aside`s are not wrapped in separate doc-fragments for the epub that motivated this issue. But if it is the exception to the rule, it might not be worth the effort to support this; I can always write a script to fix the book…

**Comment by poire-z** [comment]
I'll be looking into, It _might_ just fit in the way we handle non-linear stuff without being too ugly (if aside type=footnote, if not already in a non-linear sequence, create one - when leaving it, check if there is a next sibling and it if is also aside type=footnote, do nothing, otherwise close the non-linear sequence).

> I can always write a script to fix the book…

How would you fix it if they are not wrapped in separate doc fragments ?

**Comment by Jules-Bertholet** [comment]
> How would you fix it if they are not wrapped in separate doc fragments ?

By adding new doc fragments, moving the `aside`s into them, and fixing the links.

> And it is each footnote that is wrapped in a ASIDE, so in a book with 1000 footnotes, that makes 1000 non-linear flows + 1 linear flow, so 1001 things for the frontend, and it would look ugly on the UI side (and possibly slow).

This could be a concern for the book I am dealing with: it has around 50000 footnotes, with the `aside`s all immediately following the `a` that links to them (so none of the `aside`s are siblings).

**Comment by poire-z** [comment]
(You must have quoted the wrong text of mine :)

Indeed, this breaks my idea.
Is it the A wrapping the ASIDE ?
Or rather inserted between ASIDE like:
```html
<a id="ezhiuea"/>
<aside> <p>rearreza </p> </aside>
<a id="ezreavf"/>
<aside> <p>rezara reza </p> </aside>
```
Both feel like bad practice/HTML, a A wrapping a whole bunch of block elements, or marking a point between footnotes, instead of pointing to the container of the thing.
Does iBooks handle that case well ?!
Does our in-page footnotes, and/or popup footnotes handle that case well ?!

**Comment by poire-z** [comment]
You mean the ASIDE are inline inside the main content ?!
Do in-page footnotes work with that ?

Well, that seems impossible to solve the way we do it, and the solution above would break that main content into many ridiculous slices.
In that case, the only solution for you with this book would be a style tweak to do:
`aside[type="footnote"] { display: none; }`
and may be footnote popups would still work on them?

**Comment by Jules-Bertholet** [comment]
>In that case, the only solution for you with this book would be a style tweak to do:
aside[type="footnote"] { display: none; }
and may be footnote popups would still work on them?

This works well enough. The experience is not ideal (popup content doesn't correspond exactly to footnotes, links inside footnotes can't be clicked) but it makes the book usable. Thanks!

**Comment by poire-z** [comment]
> I'll be looking into, It might just fit in the way we handle non-linear stuff without being too ugly (if aside type=footnote, if not already in a non-linear sequence, create one - when leaving it, check if there is a next sibling and it if is also aside type=footnote, do nothing, otherwise close the non-linear sequence).

Got another idea: introduce 2 additional -cr-hint:
```c
// To be set on a block element: this block makes a non-linear flow that can be hidden from normal flow by frontends.
// Consecutive siblings all hinted With 'non-linear-combining' combine all into a single flow.
#define CSS_CR_HINT_NON_LINEAR              0x00100000 // -cr-hint: non-linear
#define CSS_CR_HINT_NON_LINEAR_COMBINING    0x00300000 // -cr-hint: non-linear-combining  (2 bits flag, includes previous one)
```
And replace the current hardcoded handling of DocFragment with NonLinear attribute, and FB2 2nd++ BODYs with some code handling these hints instead.

We would then add into our User Agent stylesheets:
```css
/* In cr3gui/data/epub.css and html5.css : */
/* Non-linear fragments (hideable) */
DocFragment[NonLinear] {
    -cr-hint: non-linear;
}
section[type=footnotes] {
    -cr-hint: non-linear;
}
aside[type=footnote] {
    -cr-hint: non-linear-combining;
}

/* In cr3gui/data/fb2.css : */
/* Mark BODYs others than the first as non-linear, so they can be hidden by frontend code.
   (this is actually suggested by the FB2 specs: "... multiple bodies are used for additional information,
   like footnotes, that do not appear in the main book flow. The first body is presented to the reader by
   default, and content in the other bodies should be accessible by hyperlinks.") */
body + body {
    -cr-hint: non-linear;
}
```

The code would avoid nested non-linear sequences, so the few books that don't set non-linear in the OPF, but use `<aside epub:type="footnote">`  with or without a  `<section epub:type="footnotes">` container could advertise the existence of non-linear flows, and frontend would show the menu item, allowing these to be hidden
Users fond of hiding stuff could even set it on non-epub:type-enabled footnotes or footnotes containers in a style tweak.
It it becomes crappy (a new flow and so a page-break for each) because some intermediate crap between `<aside>`, or because the `<aside>` are after each paragraph they occur in instead of at end of chapter, then just don't enable it.
If some things get crappy, or users don't want to see any "Hide non-linear flows", they can use in a style tweak `-cr-hint: none` for some of these targets.

In the case there is no proper wrapper for all footnotes, one could still see a page in the main flow with the heading "Notes:", or some trailing `<HR>` in the main flow. This make it not really usable with Wikipedia EPUBs if we use `ol.references { -cr-hint: non-linear; }` as the headings are outside all OL>LI wrappers:

![image](https://user-images.githubusercontent.com/24273478/149621419-4399b53e-3631-466a-9008-66749573c200.png)


This would not be triggered on `<aside>` that are "display: inline", so the above trick of making them `display: none`, and only use footnote popups, would still be needed.

This is orthogonal to "in-page footnotes", as one may want to hide or keep them in the normal flow whether he is using "in-page" or "popup" footnotes.

@Jellby : any thought about this and me making these changes to your https://github.com/koreader/crengine/pull/391 (only the bits in lvrend.cpp) ?

**Comment by Jellby** [comment]
> @Jellby : any thought about this and me making these changes to your [koreader/crengine#391](https://github.com/koreader/crengine/pull/391) (only the bits in lvrend.cpp) ?

I have zero experience with `<aside>` (or ePub3 in general), so I can't say about the implications. But if it doesn't break the handling of ePub2 non-linear flows, I have no objection :)

**Comment by poire-z** [comment]
Using some random issue to remember (and get a mail) for some issue with using `-cr-hint: non-linear`:
If set on something like `<div class=calibre15>` (the container of footnotes):
`div.calibre15 {-cr-hint: non-linear;}`
If in the book there are other occurences of it with empty content, ie.:
`<div class="calibre15"><a class="calibre2"></a></div>`
these will be counted as a flow - but having no content, it will somehow not be involved in page splitting.
But then, if we have 3 such empty div at the start of book, the 4th (the 1st of interest) will have flow number 4. And in frontend, we will get the array of flows with 3 `nil` at start, and frontend code use  `self.flows[1] ~= nil` to detect if a document has non-linear flows, and will detect it has not.

This is to be fixed. Dunno yet how (crengine not counting worthless ones, or frontend checking other things?).
Workaround for my current book (hint for my future self fixing it: Castillo/Idiot): checking if these div have content:
`div.calibre15[_] {-cr-hint: non-linear;}`
